gpflow/kernels/categorical.py

```diff
@@ -0,0 +1,97 @@
+from typing import Any, Optional
+
+import numpy as np
+import tensorflow as tf
+import tensorflow_probability as tfp
+
+from .. import default_int, set_trainable
+from ..base import Parameter, TensorType
+from . import Kernel
+
+tfd = tfp.distributions
+st = tfp.bijectors.Softplus().inverse
+
+
+def latent_from_labels(Z: tf.Tensor, labels: TensorType) -> TensorType:
+    """
+    Converts a tensor of labels into a tensor of the latent space values of those labels.
+    :param Z: Z is a tensor whose entries are the latent space values of the relevant label
+        i.e. Z[3] is the value of label 3.
+        [num_labels]
+    :param X: X is the input data.  The last column should contain the labels as integers.
+        [..., N, D+1]
+    :return: [..., N, D + label_dim]
+    """
+    indices = tf.cast(labels, default_int())  # [..., N]
+    return tf.gather(Z, indices)  # [..., N, label_dim]
+
+
+def _concat_inputs_with_latents(Z: tf.Tensor, X: TensorType) -> TensorType:
+    """
+    Transforms a dataset X from category space into the latent space.
+    :param Z: Z is a tensor whose entries are the latent space values of the relevant label
+        i.e. Z[3] is the value of label 3.
+        [num_labels]
+    :param X: X is the input data.  The last column should contain the labels as integers.
+        [..., N, D+1]
+    :return: [..., N, D + label_dim]
+    """
+    labels = X[..., -1]
+    latent_values = latent_from_labels(Z, labels)
+
+    return tf.concat([X[..., :-1], latent_values], axis=-1)  # [..., N, x_dim + label_dim]
```

Comment:
```suggestion
    return tf.concat([X[..., :-1], latent_values], axis=-1)  # [..., N, D + label_dim]
```

Also maybe use `@check_shapes`?